test/e2e/suites/e2e/e2e_test.go

```diff
@@ -569,33 +570,51 @@ var _ = Describe("e2e tests [PR-Blocking]", func() {
 					Equal(*machine.Spec.FailureDomain),
 					fmt.Sprintf("Server %s was not scheduled in the correct AZ", machine.Name))
 
-				// Check that all machines have the expected boot volume
+				// Check that all machines have the expected volumes:
+				// - 1 root volume
+				// - 1 additional volume
 				volumes := server.AttachedVolumes
-				Expect(volumes).To(HaveLen(1))
+				Expect(volumes).To(HaveLen(2))
 
-				bootVolume, err := shared.GetOpenStackVolume(e2eCtx, volumes[0].ID)
+				rootVolume, err := shared.GetOpenStackVolume(e2eCtx, volumes[0].ID)
```

Comment:
it's apparently stable. The volume attachments listed for a server seem to be indexed by their creation, and the rootVolume is always created first. I was surprised myself too but I've tested it a few times and it was always the case :shrug: .

I'll try to find the code in Nova to confirm what I saw.

Reply:
The API output is generated [here](https://github.com/openstack/nova/blob/27.1.0/nova/api/openstack/compute/views/servers.py#L730-L749). The `bdms` parameter is populated [here](https://github.com/openstack/nova/blob/27.1.0/nova/api/openstack/compute/views/servers.py#L365-L370). I followed that call to `nova.objects.BlockDeviceMappingList.bdms_by_instance_uuid` all the way down to the DB layer and didn't spot any sorting being done. This leads me to believe we're probably relying on the order of the rows being returned by the DB. As you likely know, MySQL doesn't guarantee sorting without a specific `ORDER BY` so that doesn't seem wise since it won't be consistent.

You should double check this (!) but assuming this is correct, you probably want to explore device tagging. In general, nova doesn't guarantee anything wrt device ordering and asks that you use [device tagging](https://specs.openstack.org/openstack/nova-specs/specs/mitaka/approved/virt-device-role-tagging.html) if you need this. I don't know if gophercloud support this but you simply want to add the `tags` field to the BDM for the root volume, e.g.

```json
{
  "block_device_mapping_v2": [
    {
      "uuid": "<image_id>",
      "boot_index": 0,
      "source_type": "image",
      "destination_type": "volume",
      "volume_size": "<volume_size>",
      "tags": ["foo", "bar"]
    }
  ]
}
```

(note: this requires API microversion 2.42).

You can then use the volume attachments API (`GET /servers/{serverID}/os-volume-attachments`) to list the attached volumes with tags, since this info is not returned in the `os-extended-volumes:volumes_attached` field in the `GET /servers/{serverID}` response.

Reply:
right, that was my thought as well. I'll investigate it.

Reply:
For now, I've addressed this comment with:

https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/6855f55964b2a57eddddafc74af6502fb389c5ee/test/e2e/suites/e2e/e2e_test.go#L579-L592

Until we find a better solution. With device tagging that I plan to work on next, we'll probably make this better.